# Fonctions

Inscrivez tout d'abord <span style="color: red">CLARY Emilie, DIEU Joachim</span>.

Nous allons dans cette séance de travaux pratiques manipuler des fonctions entre ensembles finis, qu'on peut implémenter à l'aide de tableaux associatifs, appelés <a href="https://docs.python.org/2/tutorial/datastructures.html#dictionaries">dictionnaires</a> en Python (l'équivalent des maps de C++). Ici une fonction de l'ensemble $\{ a, b, c, d\}$ vers $\{1,2,3\}$:

In [13]:
f = { "a": 1, "b": 2, "c": 7, "d": 3 }

On évalue la fonction sur un élément du domaine à l'aide de l'opérateur crochets:

In [14]:
f["b"]

2

Voici une fonction utilitaire qui permet d'afficher sur deux lignes la description en extension d'une fonction : $y$ en-dessous de $x$ signifie que $f(x) = y$.

In [15]:
def two_line_repr(f):
    
    res = "[ "
    for x in f.keys():
        pad = (len(str(f[x])) - len(str(x))) * " " if len(str(f[x])) > len(str(x)) else ""
        res += pad + str(x) + " "
    res += "]\n[ "
    
    for x in f.keys():
        pad = (len(str(x)) - len(str(f[x]))) * " " if len(str(x)) > len(str(f[x])) else ""
        res += pad + str(f[x]) + " "
    res += "]"
    
    print(res)

In [16]:
two_line_repr(f)

[ a b c d ]
[ 1 2 7 3 ]


## 1) Composition

Définir une fonction **comp(f,g)** qui prend en argument deux fonctions $f$ et $g$ données sous forme de dictionnaire et renvoie la fonction composée $f \circ g$; une erreur est affichée si les fonctions ne sont pas composables. Vérifier sur quelques exemples que vous avez bien le fonctionnement attendu.

In [17]:
def comp(f,g) :
    return { x : f[g[x]] for x in g.keys()}

In [18]:
g = {i: 3*i%20 for i in range(20)}
two_line_repr(g)

[ 0 1 2 3  4  5  6 7 8 9 10 11 12 13 14 15 16 17 18 19 ]
[ 0 3 6 9 12 15 18 1 4 7 10 13 16 19  2  5  8 11 14 17 ]


In [19]:
d = [10, 11, 12, 9, 8, 12, 1, 10, 2, 6, 9, 7]

f = {i+1: d[i] for i in range(12)}

r = f

for i in range(7) :
    
    r = comp(r, f)

two_line_repr(r)

[  1 2 3 4 5 6 7  8  9 10 11 12 ]
[ 12 9 1 2 6 1 6 12 11  7  2 10 ]


## 2) Générer les fonctions

Définir une fonction (récursive si vous voulez) **all_func(n)** prenant en argument un entier $n$ et renvoyant une liste de toutes les fonctions de l'ensemble $[\![1, n]\!]$ dans lui-même (vérifiez que vous en obtenez bien $n^n$).

In [20]:
def all_func(n):
    return all_func_mn(n,n)

def all_func_mn(m,n):
    list = []

    if n == 0:
        return [{}]

    for i in all_func_mn(m,n-1):
        
        for j in range(1,m+1):
            
            tmp = i.copy();
            tmp[n] = j
            list.append(tmp)

    return list;

print(all_func(3))
print("Nombre de fonctions dans l'ensemble [1,n] = %d " % len(all_func(3)))

[{1: 1, 2: 1, 3: 1}, {1: 1, 2: 1, 3: 2}, {1: 1, 2: 1, 3: 3}, {1: 1, 2: 2, 3: 1}, {1: 1, 2: 2, 3: 2}, {1: 1, 2: 2, 3: 3}, {1: 1, 2: 3, 3: 1}, {1: 1, 2: 3, 3: 2}, {1: 1, 2: 3, 3: 3}, {1: 2, 2: 1, 3: 1}, {1: 2, 2: 1, 3: 2}, {1: 2, 2: 1, 3: 3}, {1: 2, 2: 2, 3: 1}, {1: 2, 2: 2, 3: 2}, {1: 2, 2: 2, 3: 3}, {1: 2, 2: 3, 3: 1}, {1: 2, 2: 3, 3: 2}, {1: 2, 2: 3, 3: 3}, {1: 3, 2: 1, 3: 1}, {1: 3, 2: 1, 3: 2}, {1: 3, 2: 1, 3: 3}, {1: 3, 2: 2, 3: 1}, {1: 3, 2: 2, 3: 2}, {1: 3, 2: 2, 3: 3}, {1: 3, 2: 3, 3: 1}, {1: 3, 2: 3, 3: 2}, {1: 3, 2: 3, 3: 3}]
Nombre de fonctions dans l'ensemble [1,n] = 27 


## 3) Associativité

Vérifier explicitement l'associativité de la composition sur $\mathcal{F}_n :=\{ f : [\![1,n]\!] \to [\![1,n]\!] \}$ pour $n=3$ (rapide) puis pour $n = 4$ ($\sim$ 1 min 20 s sur ma machine &ndash; vous pouvez chronométrer l'exécution d'une cellule en y écrivant **%%time** sur la première ligne). Combien de temps cela vous prendrait-il pour $n = 5$ ?

In [21]:
def is_func_associatif(n):
    
    F = all_func(n)
    nbr = 0
    asso = 0

    for i in F :
        
        for j in F :
            
            for k in F :

                nbr +=1
                
                if comp(comp(i,j),k) == comp(i,comp(j,k)): # On teste si  ( i ∘ j ) ∘ k  =  i ∘ ( j ∘ k )
                    asso +=1
    
    return asso == nbr

In [49]:
%%time

F = all_func(3)

show(associatif(3))

True

CPU times: user 118 ms, sys: 2 µs, total: 118 ms
Wall time: 117 ms


In [58]:
%%time

F = all_func(4)

show(associatif(4))

True

CPU times: user 1min 40s, sys: 6.55 ms, total: 1min 40s
Wall time: 1min 41s



On fait le rapport entre all_func(4) et all_func(3) : 101/0,117 = 863,25.

all_func(4) x 863,25 = 87 000 secondes = 24h environ.


## 4) Fonctions symétrisables

Déterminer par force brute le nombre de fonctions $f \in \mathcal{F}_n$ qui sont symétrisables pour $\circ$ pour  $n \leq 5$. Ces valeurs correspondent-elles à ce que vous attendiez ?

In [63]:
%%time


for i in range(1,6):
    
    nbr = 0
    
    for j in all_func(i):
        
        for k in all_func(i):
            
            F = comp(j,k)
            try_ = 0
            
            for l in range(1,i+1):
                
                if F[l] == l : # Si on est en présence du neutre,
                    try_ +=1 # On ajoute 1 au compteur
            
            if try_ == i:
                nbr+=1
    
    show("%d : % d"%(i,nbr))
    

'1 :  1'

'2 :  2'

'3 :  6'

'4 :  24'

'5 :  120'

CPU times: user 42.4 s, sys: 6.65 ms, total: 42.4 s
Wall time: 42.5 s


Les fonctions symétrisables sont bijectives, or dans un groupe il y a n! fonctions bijectives donc n! fonctions symétrisables.

1! = 1 = 1
2! = 2x1 = 2
3! = 3x2x1 = 6
4! = 4x3x2x1 = 24
5! = 5x4x3x2x1 = 120

6! = 6x5x4x3x2x1 = 720

Il y aurait donc 720 fonctions symétrisables pour n = 6.

## 5) Fonctions qui commutent

Déterminer le nombre de paires $(f,g) \in \mathcal{F}_n^2$ de fonctions qui commutent pour $n \leq 5$ (30 s sur mon PC). Sauriez-vous dire combien il y en a pour $n = 6$ ? (vous pouvez <a href="https://oeis.org/">vous aider des internets</a> si vous séchez)

In [62]:
%time

for i in range(1,6):
    
    nbr = 0
    
    for j in all_func(i): # On parcourt les lignes
        
        for k in all_func(i): # et les colonnes
            
            if comp(j,k) == comp(k,j): # On teste si j ∘ k = k ∘ j
                nbr +=1 # Si oui on rajoute 1
                
    show("%d : %d"%(i,nbr))


CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 7.87 µs


'1 : 1'

'2 : 10'

'3 : 141'

'4 : 2824'

'5 : 71565'

Par rapport, on peut extrapoler qu'il y aurait au moins 30 fois 71 565 fonctions commutatives pour n = 6. Soit au moins 2146950.

## 6) Orbites

On définit l'*orbite* d'un élément $x \in [\![1,n]\!]$ par une fonction $f \in \mathcal{F}_n$ comme étant l'ensemble des itérés de la fonction sur cet élément:

$$ \mathrm{Orb}_f(x) = \{ x, \, f(x), \, f^2(x), \, f^3(x), \, \ldots \, \} $$

Définir une fonction **orb(f,x)** renvoyant l'ensemble de ces valeurs sous forme de liste. Attention: l'ensemble d'arrivée de $f$ étant fini, la suite des valeurs $(f^n(x))$ finira forcément par boucler; assurez-vous de le détecter et de vous arrêter à ce moment.

In [66]:
def orb(f,x):
    
    orbit = []
    orbit.append(x);
    fn = f
    
    while fn[] in orbit:
        
        orbit.append(fn[])
        fn = comp(fn, f)
        
    return orbit 

# Ne marche pas

SyntaxError: invalid syntax (3051984716.py, line 7)

Quelle est la taille moyenne des orbites des fonctions $f \in \mathcal{F}_7$ ? i.e. évaluez
$$ \frac{1}{|\mathcal{F}_7|} \sum_{f \in \mathcal{F}_7} \frac{1}{7} \sum_{x = 1}^7 | \mathrm{Orb}_f(x) | $$

In [ ]:
%%time

moy = 0

for i in all_func(7):
    sum = 0
    for j in range(1,8):
        sum += len(orb(i,j))   # On ajoute chaque valeur,     
    moy += sum/7 # Puis on divise par le nombre de valeurs

print(moy/(7**7))